### DISCUSSION TITLE
Using Progress inside Panels

### DISCUSSION THREAD

**Discussion by PoshoDev** [Textualize/rich#2680] [Q&A]
Hey there, 🙂

I'm trying to draw some progress bars inside of Rich Panels. I'm reading throughout the docs and forums, but the only implementation I see is for Stdout only. Is there a way to do this? This is more or less what I want to do:

![lol](https://i.imgur.com/MvjKMLd.png)

Alternatively, I already attempted doing my own progress bars but the issue is that Rich Panels' widths are always `None` unless they're explicitly defined, which isn't always the case, so doing this becomes trickier. Any advice?

Thank you for reading!

Cheers,
Posho ✌

**Comment by geekya215** [comment]
> Alternatively, I already attempted doing my own progress bars but the issue is that Rich Panels' widths are always None unless they're explicitly defined, which isn't always the case, so doing this becomes trickier. Any advice?

I think you should use `Panel.fit`, as the rich document say **Panels will extend to the full width of the terminal. You can make panel fit the content by setting `expand=False` on the constructor, or by creating the Panel with `fit`.**

the following code will satisfy your requirements. You can customize the functionality based on the code given below, I suggest you read [Progress](https://rich.readthedocs.io/en/stable/progress.html) and [Live](https://rich.readthedocs.io/en/stable/live.html) again, these sections of the documentation will be helpful before you start.

hope my answer can solve your problem :smiley: 
```python
import time

from rich.live import Live
from rich.panel import Panel
from rich.progress import (
    Progress,
    TimeRemainingColumn,
    TaskProgressColumn,
    TextColumn,
    BarColumn,
    SpinnerColumn,
)

progress = Progress(
    TextColumn("[progress.description]{task.description}"),
    SpinnerColumn(),
    BarColumn(),
    TaskProgressColumn(),
    TimeRemainingColumn())

task1 = progress.add_task("[red]Downloading...", total=1000)
task2 = progress.add_task("[green]Processing...", total=1000)
task3 = progress.add_task("[cyan]Cooking...", total=1000)

panel = Panel.fit(progress, title="Title", subtitle="Footer")

with Live(panel, refresh_per_second=4) as live:
    while not progress.finished:
        progress.update(task1, advance=0.5)
        progress.update(task2, advance=0.3)
        progress.update(task3, advance=0.9)
        time.sleep(0.02)
```
![rich_progressbar_with_panel](https://user-images.githubusercontent.com/10865729/205433572-55c2d95a-4c87-4a36-8126-dbf91904b3fb.png)
